File: .claude/hooks/worktree-guard.mjs
#!/usr/bin/env node
// @hookstack pre-edit-worktree-guard
// Empêche l'édition de fichiers hors du worktree courant (PreToolUse Write|Edit)
import { readFileSync } from 'fs';
import { execSync } from 'child_process';
import { resolve } from 'path';
import { homedir } from 'os';
import { fileURLToPath } from 'url';

// Répertoires internes des agents — légitimes hors worktree
const AGENT_DIRS = ['.claude', '.codex'];

function defaultExec(cmd) {
  return execSync(cmd, { encoding: 'utf8', timeout: 5_000 }).trim();
}

export function run(input, { exec = defaultExec, home = homedir() } = {}) {
  const filePath = input.tool_input?.file_path ?? '';
  if (!filePath) return null;

  try {
    const worktreeRoot = exec('git rev-parse --show-toplevel');
    const worktreeLines = exec('git worktree list').split('\n');
    const mainRoot = worktreeLines[0]?.split(/\s+/)[0] ?? '';

    // N'applique le garde que dans un worktree non principal
    if (!mainRoot || worktreeRoot === mainRoot) return null;

    const absFile = resolve(filePath);

    // Autoriser les répertoires internes des agents (plans, mémoire, config…)
    if (AGENT_DIRS.some((d) => absFile.startsWith(resolve(home, d) + '/'))) return null;

    if (!absFile.startsWith(worktreeRoot + '/')) {
      return {
        decision: 'block',
        reason: `Écriture hors du worktree courant (${worktreeRoot}). Vérifiez le chemin cible.`,
      };
    }
  } catch {
    // git absent ou pas dans un repo — laisser passer
  }
  return null;
}

/* v8 ignore next 5 */
if (process.argv[1] === fileURLToPath(import.meta.url)) {
  const input = JSON.parse(readFileSync(0, 'utf8'));
  const result = run(input);
  if (result) process.stdout.write(JSON.stringify(result));
}

File: .claude/hooks/worktree-guard.py
#!/usr/bin/env python3
# @hookstack pre-edit-worktree-guard
"""Blocks edits outside the current worktree (PreToolUse Write|Edit)."""
import json
import os
import subprocess
import sys

# Répertoires internes des agents — légitimes hors worktree
AGENT_DIRS = [".claude", ".codex"]


def _default_exec(cmd):
    return subprocess.run(
        cmd, shell=True, capture_output=True, text=True, timeout=5, check=True
    ).stdout.strip()


def run(input_data=None, *, exec_cmd=None, home=None):
    if exec_cmd is None:
        exec_cmd = _default_exec
    if home is None:
        home = os.path.expanduser("~")

    data = input_data or {}
    file_path = (data.get("tool_input") or {}).get("file_path") or ""
    if not file_path:
        return None

    try:
        worktree_root = exec_cmd("git rev-parse --show-toplevel")
        worktree_lines = exec_cmd("git worktree list").split("\n")
        main_root = worktree_lines[0].split()[0] if worktree_lines and worktree_lines[0] else ""

        # N'applique le garde que dans un worktree non principal
        if not main_root or worktree_root == main_root:
            return None

        abs_file = os.path.abspath(file_path)

        # Autoriser les répertoires internes des agents (plans, mémoire, config…)
        if any(abs_file.startswith(os.path.join(home, d) + os.sep) for d in AGENT_DIRS):
            return None

        if not abs_file.startswith(worktree_root + os.sep):
            return {
                "decision": "block",
                "reason": f"Écriture hors du worktree courant ({worktree_root}). Vérifiez le chemin cible.",
            }
    except Exception:
        # git absent ou pas dans un repo — laisser passer
        pass
    return None


if __name__ == "__main__":
    data = json.load(sys.stdin)
    result = run(data)
    if result:
        sys.stdout.write(json.dumps(result))
